compiler/rustc_mir/src/transform/instrument_coverage.rs

```diff
@@ -303,86 +652,394 @@ impl<'a, 'tcx> Instrumentor<'a, 'tcx> {
+                        for dup in pending_dups.drain(..) {
+                            debug!("    ...adding at least one pending={:?}", dup);
+                            coverage_spans.push(dup);
                         }
-                        CoverageKind::Unreachable => {
-                            if !INCLUDE_COVERAGE_STATEMENTS {
-                                continue;
-                            }
-                            String::from("unreachable")
+                    } else {
+                        pending_dups.clear();
+                    }
+                }
+            }
+            debug!("FOR curr={:?}", curr);
+            let curr_span = curr.span;
+            if prev.span.lo() > curr.span.lo() {
+                // This can only happen if a prior iteration updates `prev` to skip forward, such as
+                // skipping past a closure.
+                debug!(
+                    "  prev.span starts after curr.span, so curr will be dropped (skipping past \
+                    closure?); prev={:?}",
+                    prev
+                );
+                continue; // without changing prev
+            } else if curr.is_mergeable(&prev) {
+                debug!("  same bcb (and neither is a closure), merge with prev={:?}", prev);
+                curr.merge_from(prev); // Note that curr.span may now differ from curr_span
+            } else if prev.span.hi() <= curr.span.lo() {
+                // different and disjoint spans, so keep both
+                debug!(
+                    "  different bcbs and disjoint spans, so keep curr for next iter, and add \
+                    prev={:?}",
+                    prev
+                );
+                coverage_spans.push(prev);
+            } else if prev.is_closure {
+                // drop any equal or overlapping span (`curr`) and keep `prev` to test again in the
+                // next iter
+                debug!(
+                    "  curr overlaps a closure (prev). Drop curr and keep prev for next iter. \
+                    prev={:?}",
+                    prev
+                );
+                continue;
+            } else if curr.is_closure {
+                // if `prev`s span extends left of the closure (`curr`), carve out the closure's
+                // span from `prev`'s span. (The closure's coverage counters will be injected when
+                // processing the closure's own MIR.) Add the portion of the span to the left of the
+                // closure; and if the span extends to the right of the closure, update `prev` to
+                // that portion of the span. For any `pending_dups`, repeat the same process.
+                let left_cutoff = curr.span.lo();
+                let right_cutoff = curr.span.hi();
+                let has_pre_closure_span = prev.span.lo() < right_cutoff;
+                let has_post_closure_span = prev.span.hi() > right_cutoff;
+                if has_pre_closure_span {
```

Comment:
I noticed that right around here my eyes start glossing over a bit. If you have any ideas for how to simplify this nested logic tree, I think it would help. Maybe extracting helper functions so it's all high level code.

Reply:
I've been refactoring this today. Update coming shortly. Thanks.

Reply:
Done.

@tmandry - I refactored the `coverage_span()` logic into a new struct/impl `CoverageSpanRefinery`. Let me know if this works for you.